### PR TITLE
Prevent JIT from overspecializing to every single size configuration

### PR DESCRIPTION
**Pull request by apaszke**
Please review the expects carefully to make sure there are no regressions. I tried to go over them one by one when they changed, but it's sometimes easy to miss finer details.

Summary of changes:

- Renamed `TensorType` to `CompleteTensorType`. Added a new `TensorType` which records only the scalar type, number of dimensions, and device of a value. The argument behind the rename is to encourage people to use `CompleteTensorType` less, as most passes will only have limited information available. To make transition easier `complete_type->cast<TensorType>()` works, and makes our passes work with both kinds of specialization if they don't need extra the extra detail.
- Renamed `ArgumentSpec` to `CompleteArgumentSpec`. Added a new `ArgumentSpec`, which matches argument only at the level of the new `TensorType`.
- Shape analysis can process graphs with both `CompleteTensorType` and `TensorType`.
- Fuser was a part that heavily relied on full shape information being available. Now, we simply try to fuse the largest possible graphs, and have to do run-time checks to make sure they match the code we generate. If they don't, we fall back to regular interpretation. The shape checks are implementing using an optimized method exploiting algebraic properties of shapes with broadcasting, and the relations of broadcasting with pointwise ops. A full written proof of correctness of the shape checking algorithm is included in a comment in `graph_fuser.cpp`.

@zdevito @ezyang @mruberry @ngimel @csarofeen

### CODE REVIEW

File: test/expect/TestJit.test_broadcast_fusion_cuda.expect [hunk 1]
```diff
@@ -1,18 +1,14 @@
-graph(%0 : Float(4, 4)
-      %1 : Float(4)
-      %2 : Float(4)) {
-  %3 : int[] = prim::Constant[value=[4, 4]]()
-  %4 : int = prim::Constant[value=0]()
-  %5 : Float(4!, 4) = aten::expand(%1, %3, %4)
-  %6 : Float(4!, 4) = aten::expand(%2, %3, %4)
-  %7 : Float(4, 4) = prim::FusionGroup_0[device=0](%6, %0, %5)
-  return (%7);
+graph(%0 : Float(*, *)
```
Comment (ezyang) [inline test/expect/TestJit.test_broadcast_fusion_cuda.expect]: Maybe someday we'll get a symbolic shape analysis :>

File: torch/csrc/jit/fusion_compiler.cpp [hunk 2]
```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+//          operation of broadcasting (returning bottom upon shape mismatch) forms a monoid.
+//          In simpler terms: broadcasting is associative, i.e. B(a, B(b, c)) == B(B(a, b), c).
+//
+// Proof.   Satisfies all monoid laws:
+//            - Closed under broadcasting (trivial)
+//            - Empty shape is the identity element: B(a, []) == B([], a) == a
+//            - Associativity: A simple visual proof is that you can expand 3 tensors
+//                at the same time by stacking their sizes (with alignment to the right),
+//                just as you'd do in the case of 2 tensors, but with an intermediate
+//                (the algorithm ends up being pretty much the same).
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: If this were an undergrad inductive proofs class I would fail this proof :P I agree the lemma is true though.

File: torch/csrc/jit/type.h [hunk 3]
```diff
@@ -183,22 +248,22 @@ struct TORCH_API TensorType : public Type {
   }
   static TypePtr fromNumberType(TypePtr typ);
 
+  static CompleteTensorTypePtr sliceSubtypes(const CompleteTensorTypePtr& type) {
```
Comment (ezyang) [inline torch/csrc/jit/type.h]: What does this do? When would I use it?

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 4]
```diff
@@ -63,39 +63,30 @@ IValue representativeValue(Value* v) {
 
 void PropagateShapeOnBlock(Block * block, bool insert_expands=true);
 
-// for each node in the schema with type Tensor, extract the TensorType
+// for each node in the schema with type Tensor, extract the CompleteTensorType
 // returns at::nullopt if any Tensor in the schema does not have a known shape
 // ignores non-tensor in the list of inputs
```
Comment (ezyang) [inline torch/csrc/jit/passes/shape_analysis.cpp]: Isn't this out of date now?

File: torch/csrc/jit/type.h [hunk 5]
```diff
@@ -117,57 +116,123 @@ using TensorTypePtr = std::shared_ptr<TensorType>;
 // This node represents a single Tensor value with a specific size
```
Comment (ezyang) [inline torch/csrc/jit/type.h]: This is out of date now, irght?

Comment (ezyang) [reply torch/csrc/jit/type.h]: I think in general we need a more detailed description of the difference between TensorType/CompleteTensorType; which one to use in what situation; how to test if you have one or the other.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 6]
```diff
@@ -49,6 +55,82 @@ std::vector<bool> TensorDesc::findContiguous(
   return cont;
 }
 
+// Descriptor for chunk-ing an input tensor into subtensors
+// OR concat-ing an output tensor from subtensors
+struct PartitionDesc {
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: Move only?

File: torch/csrc/jit/fusion_compiler.h [hunk 7]
```diff
@@ -28,7 +36,7 @@ struct TensorDesc {
   : TensorDesc(type, TensorDesc::findContiguous(sizes, strides)) {}
   TensorDesc(const at::Tensor& t)
     : TensorDesc(t.type().scalarType(), t.sizes(), t.strides()) {}
-  TensorDesc(TensorTypePtr type)
+  TensorDesc(CompleteTensorTypePtr type)
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.h]: This constructor is pretty weird. Aren't you always constructing TensorDesc from real tensors, because you want to actually run the fusion kernel?

File: torch/csrc/jit/fusion_compiler.h [hunk 8]
```diff
@@ -45,111 +53,84 @@ struct TensorDesc {
...
-  // We keep these around for debugging
-  std::string compilation_unit;
-  std::vector<TensorDesc> input_desc;
-  std::vector<TensorDesc> output_desc;
-
-  // same size as output_desc, describes whether
-  // an output is actually a concatenation of
-  // many subtensors that the fusion group produces
-  std::vector<PartitionDesc> concat_desc;
+struct FusedKernelCache {
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.h]: Needs a comment. In particular, what does it "cache" over?

File: torch/csrc/jit/fusion_compiler.cpp [hunk 9]
```diff
@@ -67,6 +149,17 @@ std::ostream& operator<<(std::ostream & out, const TensorDesc & d) {
   return out;
 }
 
+Node* usedInFusedChunk(Value * input) {
+  auto uses = input->uses();
+  if (uses.size() == 1) {
+    Node *user = uses[0].user;
+    if (user->kind() == prim::FusedChunk) {
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: What happened to the usage assert?

File: torch/csrc/jit/fusion_compiler.cpp (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Except it's not such a class, and I don't want to make the whole proof 500 LOC long, because that's not where the whole difficulty lies. It's an exercise to the reader to fill in the simple parts 😊

It basically boils down to a single dim case, where its shape takes one of the values: missing, 1, 2, 3, 4 (where 2, 3, 4 are basically to make them non-broadcastable). Then you take all combinations of 3 values from this set and show that the order of broadcasting doesn't matter. I didn't really feel like writing out the solutions for 64 cases, but you can see it is true. I can write a Python script to verify that for you haha

File: torch/csrc/jit/fusion_compiler.cpp [hunk 10]
```diff
@@ -463,19 +545,16 @@ std::tuple<std::vector<PartitionDesc>,std::vector<PartitionDesc>,bool> emitCompi
...
-
+        chunk_desc.emplace_back(agraph.input_desc[input_index++], chunks, dim);
         for (auto * o : chunk->outputs()) {
           flat_inputs.emplace_back(o, *chunk_desc.back().subtensorDesc);
         }
-        ++input_index;
       } else {
-        flat_inputs.emplace_back(p, agraph.input_desc[input_index++]);
         chunk_desc.emplace_back();
+        flat_inputs.emplace_back(p, agraph.input_desc[input_index++]);
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: If `input_index++` is always incremented, why not just move it out of the conditional XD

File: torch/csrc/jit/fusion_compiler.cpp [hunk 11]
```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+//                          (Lemma 1.)
+//                     == B(QS..., RS...)
+//                          (repeated shapes don't matter for broadcasting)
+//                     == B((QS + RS)...)
+//
+// Lemma 3. Expands are distributive over pointwise ops, i.e. E(f(a, b), s) = f(E(a, s), E(b, s))
+// Lemma 4. Expands can be collapsed, i.e. E(E(x, s1), s2) = E(x, B(s1, s2)).
+// Proof.   A simple exercise for the reader :)
+//
+// Theorem. If all (pre-concat-)outputs have equal shapes, then we can push the expands to
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: In general you can, but you can't if you want to use the fused kernel. In this case, it would fail the check and use the fallback, which can operate on arbitrary shapes.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 12]
```diff
@@ -786,32 +871,37 @@ void CompiledFusionFunction::launch_with_tensors(at::ArrayRef<at::Tensor> inputs
...
-void CompiledFusionFunction::launch(at::ArrayRef<at::Tensor> inputs, std::vector<at::Tensor> & outputs) {
+void FusedKernel::launch(at::ArrayRef<at::Tensor> inputs, std::vector<at::Tensor> & outputs) {
   at::DeviceGuard guard(inputs.back());
+  JIT_ASSERT(inputs.size() > 0);
+  auto & ref_type = inputs[0].type();
   outputs.clear();
   outputs.reserve(outputDescriptors().size());
   for(auto & od : outputDescriptors()) {
-    outputs.push_back(torch::getType(backend(),od.scalar_type).tensor());
+    outputs.push_back(ref_type.toScalarType(od.scalar_type).tensor());
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: Thanks!

File: torch/csrc/jit/type.h (hunk 3 above)
Comment (apaszke) [reply torch/csrc/jit/type.h]: I think I'll slightly change this part before we merge it. There's a subtlety in when you do `type_ptr->cast<TensorType>()`, because if it's a `CompleteTensorType`, it will give you a `TensorTypePtr`. Now, if you later do `node->output()->setType(type->expect<TensorType>())`, then your code probably assumes that you don't care about the sizes, but in reality you never sliced the derived object to the `TensorType` superclass, and so `node->output()` will also inherit its shapes and strides, which might not be valid.

I discovered this because it caused a bug in shape prop. I had a `chunk` with an input with `CompleteTensorType`, it didn't have a case in the complete-shape case, but it did fit into the simplified function, where it simply did
```cpp
std::vector<TensorTypePtr> tensor_types;
...
for (Value * output : node->outputs()) {
  output->setType(tensor_types.at(0));
}
```
thus inheriting pre-chunk sizes in the output, which is obviously incorrect.

Comment (apaszke) [reply torch/csrc/jit/type.h]: Regarding the change I want to make: I want to `->cast<TensorType>()` to always slice the object now, so we can prevent those kinds of errors in the future.

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 4 above)
Comment (apaszke) [reply torch/csrc/jit/passes/shape_analysis.cpp]: It's slightly out of date. `CompleteTensorType` should change to the name of the template parameter, but apart from that it looks good.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 6 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Move + removed the chunk constructor, because it was unnecessary.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 13]
```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
-    auto t = output->type()->expect<TensorType>();
-    agraph.output_desc.emplace_back(t);
+  return map_size;
+}
+
+void FusedKernelCache::runFallback(Stack& stack) {
+  InterpreterState(fallback_code).runOneStage(stack);
+}
+
+void FusedKernelCache::expandArgs(std::vector<at::Tensor>& args, std::vector<int64_t>& map_size) {
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: Ugh, pass by mutable reference :(

File: torch/csrc/jit/fusion_compiler.h (hunk 7 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.h]: I think it was used previously and is no longer needed. I will remove it.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 14]
```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+  auto & fn = it->second;
+
+  std::vector<at::Tensor> outputs;
+  fn->launch(args, outputs);
+  drop(stack, num_inputs);
+  pack(stack, std::move(outputs));
+}
+
+at::optional<std::vector<int64_t>> FusedKernelCache::getMapSize(at::TensorList args, at::IntList arg_subset) {
+  // NB: we leave this uninitialized, because an empty size is trivially
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: What is "this"?

File: torch/csrc/jit/fusion_compiler.cpp (hunk 9 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Writing out this code to do what it's supposed to do, *and* check the invariant would be annoying, because it would need a loop over uses. Since it's really an invariant, I don't think we need to assert it at every single step. It would be a loud failure if we were to violate it I guess.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 13 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: well I only mutate it's elements, and doing `(*args)[i]` is so awful, that I doubt the "mutable reference" rule wins here over "make the code readable".

File: torch/csrc/jit/fusion_compiler.cpp (hunk 10 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Idk, doesn't seem to be significantly improve the code if it was to be moved out of the branches.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 15]
```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+    dim_after_broadcast = std::max(dim_after_broadcast, args[arg_idx].dim());
+  }
+  std::vector<int64_t> map_size;
+  for (size_t i = 0; i < arg_subset.size(); ++i) {
+    auto & arg = args.at(arg_subset[i]);
+    auto & chunk_desc = input_chunks.at(arg_subset[i]);
+    if (chunk_desc.nSubtensors == 1) {
+      try {
+        map_size = at::infer_size(map_size, arg.sizes());
+      } catch (std::exception& e) {
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: try-catch in the hotpath makes me GRUMPY

File: torch/csrc/jit/fusion_compiler.cpp (hunk 14 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Uh this comment got moved accidentally. It was meant to be right above `map_size`.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 16]
```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+        return at::nullopt;
+      }
     } else {
-      JIT_ASSERT(canCompileOnCPU());
-      raw_func = new CPUFusionFunction(name, agraph, config_);
+      auto tensor_sizes = arg.sizes().vec();
+      int64_t num_chunks = chunk_desc.nSubtensors;
+      int64_t dim = chunk_desc.dim;
+      if (dim < 0) {
+        dim += arg.dim();
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: Shouldn't you just use the wrap dims utils here?

File: torch/csrc/jit/fusion_compiler.cpp (hunk 15 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: I could get rid of this, but that would require some extra ATen changes. We might want to do them in a next patch, especially considering the TODO above (constantly reallocating map_size).

Anyway, `try ... catch` itself doesn't cost anything if no exceptions happen, and we will completely break out of the shape checks upon the first one, so it's not like it will keep happening over and over in a tight loop.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 16 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Sure, will use that.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 13 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: If I didn't want to mutate it, it would get passed as a `const` reference, or as an `ArrayRef`

File: torch/csrc/jit/fusion_compiler.cpp [hunk 17]
```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+    auto & pdesc = input_chunks[i];
+    if (pdesc.nSubtensors == 1) {
+      if (arg.sizes().equals(map_size)) continue;
+      arg = arg.expand(map_size);
+    } else {
+      map_size.at(pdesc.dim) *= pdesc.nSubtensors;
+      if (!arg.sizes().equals(map_size)) {
+        arg = arg.expand(map_size);
+      }
+      map_size.at(pdesc.dim) /= pdesc.nSubtensors;
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: Yuck. Understandable, but yuck.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 13 above)
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: "This function is not reentrant in `map_size`"

File: torch/csrc/jit/fusion_compiler.cpp (hunk 17 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: I understand, but I really don't want to reallocate this shape for every single chunk that we have...

File: torch/csrc/jit/fusion_compiler.cpp (hunk 13 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: What does that mean exactly?

File: torch/csrc/jit/fusion_compiler.cpp [hunk 18]
```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+    if (isFloatingType(desc.scalar_type)) {
+      scalar_type = desc.scalar_type;
+      break;
+    }
   }
-  for(auto & i : outputs) {
-   agraph.output_desc.emplace_back(i);
+  JIT_ASSERT(scalar_type);
+
+  for (Value * output : graph->outputs()) {
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: What's going on here?

Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Not sure what exactly do you have in mind? It's pretty much the same code it used to be, except we can no longer use the fact that graph outputs have complete shapes filled in by shape prop.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 19]
```diff
@@ -272,15 +261,7 @@ struct GraphFuser {
...
     if (tensors_node->output()->uses().size() > 1) return false;
-    auto tensors = tensors_node->inputs();
-
-    // Our fusion code assumes that all inputs have the same shapes, so we need to check this too.
-    auto expected = tensors.at(0)->type()->cast<TensorType>();
-    if (!expected) return false;
-    return std::all_of(tensors.begin(), tensors.end(), [&expected](Value *v) {
-        auto actual = v->type()->cast<TensorType>();
-        return actual && actual->sizes() == expected->sizes();
-    });
```
Comment (ezyang) [inline torch/csrc/jit/passes/graph_fuser.cpp]: I am wondering if it is right to delete this code. Another way to do it is to do the test if the information is available, and optimistically assume that things will work out if it is not.

The diff here is a pessimization if it causes you to start fusing more operators which you shouldn't have, because the shapes aren't right. canFuseChunk is particularly dangerous.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 13 above)
Comment (ezyang) [reply torch/csrc/jit/fusion_compiler.cpp]: You'd ordinarily expect `map_size` to be read-only for a function like this but it's not. I guess this is somewhat clear from the fact that it takes a mutable reference.

File: torch/csrc/jit/passes/graph_fuser.cpp [hunk 20]
```diff
@@ -898,7 +805,7 @@ struct GraphFuser {
       Node * chunked_op = block->owningGraph()->create(producer_for_chunk_node->kind());
       chunked_op->copyAttributes(*producer_for_chunk_node);
       // Invariant: mappable operators always produce contiguous output
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Because TensorType's don't have information about contiguity, so what am I supposed to do?

**Comment by ezyang** [comment]
Question: Shouldn't `TensorType` (the non-complete version) also track contiguity? It seems that we can get pretty accurate information about it once we get inside the graph.

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 21]
```diff
@@ -205,6 +196,64 @@ bool canPropagateShapeByRunningIt(Node* node) {
   return true;
 }
 
+bool PropagateTensorShapeOnNode(
+    Node * node, bool insert_expands, std::vector<TensorTypePtr> types);
+bool PropagateCompleteShapeOnNode(
+    Node * node, bool insert_expands, std::vector<CompleteTensorTypePtr> types);
+
+void PropagateCatShape(Node * cat_node) {
```
Comment (ezyang) [inline torch/csrc/jit/passes/shape_analysis.cpp]: Is this just moved, or are there substantive changes?

**Review by ezyang — APPROVED** [review]
Nice work.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Then delete the comment

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 21 above)
Comment (apaszke) [reply torch/csrc/jit/passes/shape_analysis.cpp]: `PropagateCatShape` is new, `PropagateCompleteShapeOnNode` is mostly a move with minor tweaks (like adding a return boolean, and removing propagation by running at the end), `PropagateTensorShapeOnNode` is now.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 19 above)
Comment (ezyang) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Err, not quite, right? As in, the old code would simply choose not to consider an operator as part of the fusion group, if it didn't fulfill the necessary conditions for it to be eligible to be in the fusion group.

I can certainly believe in practice that the maximalist fusion always works. But for example, our test suite doesn't have that many "real world" programs in it, does it?

**Comment by apaszke** [comment]
We could track contiguity in `TensorShape`, but:
1. Contiguity is very tightly bound to strides, and not having to think about strides makes it much easier to work with the code, since most ops only have a size contract, but don't have a stride contract.
2. It doesn't seem to be necessary. I don't recall even a single place where it would enable us to do something we can't already do.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 18 above)
Comment (ezyang) [reply torch/csrc/jit/fusion_compiler.cpp]: I guess, I don't remember having to adjust the concat sizes in the old code. Why does this adjustment happen here?

File: torch/csrc/jit/type.h (hunk 3 above)
Comment (ezyang) [reply torch/csrc/jit/type.h]: Yeah, this is why my default assumption is that you shouldn't use subclasses in your design, unless you are ABSOLUTELY CERTAIN that you need the subtyping. In this case, just having CompleteTensorType hold a TensorType inside would have also worked.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/jit/fusion_compiler.cpp]: We had an offline conversation and I was reminded why we have this invariant and how concat works (modeled as "multiple outputs"). Would be nice if there was a comment about it somewhere...

File: torch/csrc/jit/fusion_compiler.cpp (hunk 13 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: It's true that we modify it, but it's really a local mutation, as it is restored to its original value upon return. It's really an optimization.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 2 above)
Comment (ezyang) [reply torch/csrc/jit/fusion_compiler.cpp]: I'm joking :)

File: torch/csrc/jit/fusion_compiler.cpp (hunk 18 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Because previously the adjustment would be performed by the shape prop? Now, we only know the map_size, and the concat output size is easily derivable from that.

**Comment by ezyang** [comment]
Without symbolic sizes, we can't really track stride contracts in a useful way, but ops definitely have contiguity contracts which users know about, because contiguity tells you if you can view() a tensor, e.g.

But you're right, let's add it if/when a pass actually desperately wants to know about contiguity.

File: torch/csrc/jit/passes/graph_fuser.cpp (hunk 19 above)
Comment (apaszke) [reply torch/csrc/jit/passes/graph_fuser.cpp]: Oh, my point is that testing for complete information doesn't really make sense, because it really shouldn't be available to this pass. Optimizations at the GraphExecutor level will see *only* `TensorType` and almost never `CompleteTensorType` (maybe except for some mostly constant subgraphs). 

You do have a point, but:
- It's a temporary state. We need to get this out to unblock splitting the fusers, and the deoptimization path can be easily improved in the future.
- The condition you referred to here is about inputs to concat. I'm pretty sure that the real world examples outside of RNNs will cause us much more trouble because of moving the broadcasts earlier in the graph, or simply having different numbers of elements in different parts of the graph. Irregular cats and chunks are not all that common I think.

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 21 above)
Comment (ezyang) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Thanks. I'm not so sure why we need two top-level loops for propagation. I would have expected that this high level structure would be the same in both cases, and you would just either propagated detailed or less detailed information, based on whether or not you had more or less info.

File: torch/csrc/jit/fusion_compiler.h (hunk 7 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.h]: My bad, it is still use. I create some CompleteTensorTypes manually out of `map_size`.

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 21 above)
Comment (apaszke) [reply torch/csrc/jit/passes/shape_analysis.cpp]: I'm not exactly sure what you mean. This is exactly what happens now. If all arguments have complete type information available, we will try to use the complete shape prop. If this fails, those arguments still have valid `TensorType`s, and so we'll try to use the less detailed function. Finally, if this one fails too, we try to resort to shape prop by running.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 17 above)
Comment (zdevito) [reply torch/csrc/jit/fusion_compiler.cpp]: All of the chunk and cat stuff is really messy in the fuser for 2 reasons:
1. we don't have mutable operators in the normal graph, so we cannot express the concat optimization as a rewrite that would instruct the fuser to mutate narrowed tensors.
2. Tensor allocation in chunk and cat is so expensive that slows execution time. The only reasons we do chunk inside the fuser to because we can do chunk faster if we don't need to allocate a tensor. 

We should fix these two things and then remove chunk and cat from the fuser, they are better handled elsewhere. 

In the meantime we might consider making a smaller tensor holder specific to the fuser that can do fast narrows. Then chunk and cat could be handled on the periphery of the fuser, rather than handling the PartitionDesc objects in all of these places.  Not for this PR, but something that we should consider as a refactor before doing major further feature additions.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 22]
```diff
@@ -1047,67 +1132,407 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
 
-std::shared_ptr<CompiledFusionFunction> FusionCompiler::getOrCompile(Graph & graph,
-                                                     int device,
-                                                     at::ArrayRef<at::Tensor> inputs,
-                                                     at::ArrayRef<at::Tensor> outputs) {
-  AnnotatedGraph agraph(graph, device);
-  for(auto & i : inputs) {
-   agraph.input_desc.emplace_back(i);
+  agraph.input_desc = spec.descs();
+  at::optional<at::ScalarType> scalar_type;
```
Comment (zdevito) [inline torch/csrc/jit/fusion_compiler.cpp]: This seems fragile. It is making a previously unmade assumption that the fuser won't want to mix integer and floating point types. It should at least be documented here. The other parts of codegen that cannot emit integers or mixed floats/doubles are minor and will likely be fixed by someone, and this seems like a trap when making that fix. 

Is it not possible to just take the type as-is from the output Value? If we determined it was safe to fuser (ie. it was floating point), then we checked that it was floating point, meaning the type was known.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 23]
```diff
@@ -1047,67 +1132,407 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+  std::string name = "kernel_" + std::to_string(next_kernel_id++);
+  FusedKernel * raw_func;
+  if (device != kCPUDevice) {
+#ifdef USE_CUDA
+    raw_func = new CUDAFusedKernel(name, agraph);
+#else
+    throw std::runtime_error("cannot compile a CUDA fusion group, CUDA is not enabled.");
+#endif
+  } else {
+    JIT_ASSERT(compiler.canCompileOnCPU());
```
Comment (zdevito) [inline torch/csrc/jit/fusion_compiler.cpp]: Before we release, we need to fix the fact that on Windows the CPU fuser doesn't run, and cause the GE to fail if it did a fusion. An easy workaround would have FusionCompiler just immediately go to fallback mode in this case. That way the decision whether we can do CPU fusion is localized to only the fusion compiler and not both the optimization pass and the compiler.

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 24]
```diff
@@ -252,6 +301,14 @@ void PropagateShapeOnNode(Node * node, bool insert_expands) {
       }
       return;
     }
+    // NB: We assume that all shapes are known within fused kernels
```
Comment (zdevito) [inline torch/csrc/jit/passes/shape_analysis.cpp]: This seems fragile. The primary place where you might encounter this is in someone trying to optimize the fallback pathway of a failed fusion. But that will be a case where the shapes are not all the same size. It would be better for this to be robust to whatever can be sent here.

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 25]
```diff
@@ -261,49 +318,128 @@ void PropagateShapeOnNode(Node * node, bool insert_expands) {
...
+    }
+  }
+
+  if (auto maybe_tensor_types = gatherTensorTypes<TensorType>(node)) {
+    if (PropagateTensorShapeOnNode(node, insert_expands, std::move(*maybe_tensor_types))) {
+      return;
+    }
+  }
+
+  if (canPropagateShapeByRunningIt(node))
```
Comment (zdevito) [inline torch/csrc/jit/passes/shape_analysis.cpp]: Can we just make PropagageShapeOnNodeByRunning it return a boolean like the others?

**Review by zdevito — APPROVED** [review]
This looks good -- I have a bunch of individual but minor comments and questions.  

Higher level notes for building on top of this:

* This puts the FusionCompiler at the breaking point of complexity. Further functionality added there is going require refactor some of the Chunk and Concat logic into a separate phase, given how many times we are need to check nSubtensors, and do other things to derive correct sizes. The core of the fusion is simple, but this chunk/concat stuff is getting spread all over the place.
* We may notice regressions from overhead in launching fused kernels. For reference, we know the time it takes to do chunk in the interpreter adds significant overhead, and that seems on the same order of magnitude as the extra checking added here. We will need to monitor this and optimize if necessary.

File: torch/csrc/jit/passes/batch_mm.cpp [hunk 26]
```diff
@@ -150,8 +150,8 @@ void BatchMMBlock(Block* block) {
   std::unordered_map<Node*, TreeToken> tokens;
   for (auto node : block->nodes()) {
     if (node->kind() == aten::mm &&
-        node->input(0)->type()->cast<TensorType>() &&
-        node->input(1)->type()->cast<TensorType>()) {
+        node->input(0)->type()->cast<CompleteTensorType>() &&
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/passes/batch_mm.cpp]: Yeah I've been thinking about this one, and I'm really unsure what we can do to get it to work (unless we do sth similar to fusion, but that would require a new specialized op that verifies run-time shapes).

File: torch/csrc/jit/fusion_compiler.cpp [hunk 27]
```diff
@@ -1047,67 +1132,407 @@ struct CPUFusionFunction : public CompiledFusionFunction {
...
+  for (Value * output : graph->outputs()) {
+    broadcast_groups.insert(getInputDependencies(output));
+  }
+  return std::vector<std::vector<int64_t>>{ broadcast_groups.begin(), broadcast_groups.end() };
+}
+
+void FusedKernelCache::run(Stack& stack) {
+  int64_t num_inputs = graph->inputs().size();
+  auto args = fmap(last(stack, num_inputs), [](const IValue& i) {
+                return i.toTensor();
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: Yes, but I couldn't use `last`, because `ArrayRef`s only allow for const access of members, and I would need to move them back to the stack in case of a fallback. This pass can be optimized more in the future if we'll need it, but I chose to do the simplest possible thing for now.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 22 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: > It is making a previously unmade assumption

We are currently making an assumption that the fuser only does `float` math, so I can't see how this is a problem. Having all inputs being integral types will trigger an assert failure below, so someone will likely find out about this issue on time. I can add a comment though.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 23 above)
Comment (apaszke) [reply torch/csrc/jit/fusion_compiler.cpp]: You mean this as a followup task, correct?

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 24 above)
Comment (apaszke) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Actually those cases are no longer needed, so I can just remove them.

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 25 above)
Comment (apaszke) [reply torch/csrc/jit/passes/shape_analysis.cpp]: Sure.

**Comment by apaszke** [comment]
The last commit forces slicing on types if the dynamic cast succeeds, but the kind doesn't match exactly. Normally we could simply use the copy constructor in this place, but because we're incorrectly comparing only addresses of types in many cases (because we assume they're used as singletons), the slicing needs some extra care. We really should either stop using shared pointers to types in most places, or have a subclass of shared pointer that uses equality on held elements in `operator==`.
